File: feature/shiftgen/core/scraper.py
import os
import time
from typing import List, Dict, Optional

import requests
from bs4 import BeautifulSoup
from dotenv import load_dotenv

from .config import BASE_URL


class ShiftGenScraper:
    
    def __init__(self, username: str = None, password: str = None):
        """
        Initialize the scraper with credentials.
        
        Args:
            username: Email (defaults to env variable)
            password: (defaults to env variable)
        """
        load_dotenv()
        self.username = username or os.getenv('SHIFTGEN_USERNAME')
        self.password = password or os.getenv('SHIFTGEN_PASSWORD')
        
        if not self.username or not self.password:
            raise ValueError(
                "Credentials not provided. Set SHIFTGEN_USERNAME and "
                "SHIFTGEN_PASSWORD environment variables or pass them as arguments."
            )
        
        self.session = requests.Session()
        self.logged_in = False
        self.current_site = None
        
    def login(self) -> bool:
        try:
            payload = {
                "user_session[email]": self.username,
                "user_session[password]": self.password
            }
            resp = self.session.post(f"{BASE_URL}/login", data=payload)
            
            if resp.url.endswith("/login") or "Invalid" in resp.text:
                return False
            
            self.logged_in = True
            return True
        except requests.RequestException:
            return False
    
    def navigate_to_home(self) -> bool:
        """navigate back to home pg"""
        if not self.logged_in:
            raise RuntimeError("Not logged in. Call login() first.")
        
        try:
            resp = self.session.get(f"{BASE_URL}/member/multi_site_schedule")
            return resp.status_code == 200
        except requests.RequestException:
            return False
    
    def change_site(self, site_id: str, site_name: str = "") -> bool:
        """
        Change to a different site using the dropdown.
        
        Args:
            site_id: The site ID to switch to
            site_name: Optional site name for logging
            
        Returns:
            bool: True if successful
        """
        if not self.logged_in:
            raise RuntimeError("Not logged in. Call login() first.")
        
        try:
            self.navigate_to_home()
            time.sleep(1)
            
            resp = self.session.post(
                f"{BASE_URL}/member/change_selected_site",
                data={"site_id": site_id}
            )
            
            if resp.status_code == 200:
                self.current_site = site_name or site_id
                return True
            return False
        except requests.RequestException:
            return False
    
    def navigate_to_all_schedules(self) -> bool:
        """go to all schedules"""
        if not self.logged_in:
            raise RuntimeError("Not logged in. Call login() first.")
        
        try:
            resp = self.session.get(f"{BASE_URL}/member/schedule")
            return resp.status_code == 200
        except requests.RequestException:
            return False
    
    def fetch_schedules(self) -> List[Dict[str, any]]:
        """
        Fetch all available schedules from the current site.
        
        Returns:
            List of schedule dictionaries with 'id' and 'title'
        """
        if not self.logged_in:
            raise RuntimeError("Not logged in. Call login() first.")
        
        if not self.navigate_to_all_schedules():
            return []
        
        time.sleep(1)
        
        try:
            resp = self.session.get(f"{BASE_URL}/member/schedule")
            soup = BeautifulSoup(resp.text, "html.parser")
            
            schedules = []
            for form in soup.find_all("form", {"action": "/member/schedule"}):
                sched_id_input = form.find("input", {"name": "[id]"})
                if not sched_id_input:
                    continue
                    
                sched_id = sched_id_input.get("value")
                header_elem = form.find("h2")
                if not header_elem:
                    continue
                    
                header = header_elem.get_text(strip=True)
                schedules.append({
                    "id": sched_id, 
                    "title": header,
                    "site": self.current_site
                })
            
            return schedules
        except requests.RequestException:
            return []
    
    def get_printable_schedule(self, schedule_id: str) -> Optional[str]:
        """
        Get printable version of a schedule.
        
        Args:
            schedule_id: ID of the schedule to fetch
            
        Returns:
            HTML content of the printable schedule
        """
        if not self.logged_in:
            raise RuntimeError("Not logged in. Call login() first.")
        
        try:
            resp = self.session.post(
                f"{BASE_URL}/member/schedule",
                data={"[id]": schedule_id, "commit": "Create Print Version"}
            )
            
            return resp.text if resp.status_code == 200 else None
        except requests.RequestException:
            return None

File: src/lib/shiftgen/scraper.ts
/**
 * ShiftGen Scraper
 *
 * Authenticates against ShiftGen and fetches schedule pages.
 *
 * The old "switch site -> list schedules -> POST 'Create Print Version'" flow is
 * gone. That flow depended on the legacy markup (`form[action="/member/schedule"]`
 * with an `<h2>` title), and the current pages contain no forms at all. The print
 * endpoint now returns a wkhtmltopdf-rendered PDF rather than HTML, so it can no
 * longer feed an HTML parser either.
 *
 * The multi-site schedule view replaces all of it: one authenticated GET returns
 * every schedule the account can see, with each shift's data in element attributes.
 */

import { BASE_URL, MULTI_SITE_SCHEDULE_PATH } from './config';

export class ShiftGenScraper {
  private username: string;
  private password: string;
  private cookies: Map<string, string> = new Map();
  private loggedIn: boolean = false;

  constructor(username?: string, password?: string) {
    this.username = username || process.env.SHIFTGEN_USERNAME || '';
    this.password = password || process.env.SHIFTGEN_PASSWORD || '';

    if (!this.username || !this.password) {
      throw new Error(
        'Credentials not provided. Set SHIFTGEN_USERNAME and SHIFTGEN_PASSWORD environment variables or pass them as arguments.'
      );
    }
  }

  /**
   * Login to ShiftGen
   */
  async login(): Promise<boolean> {
    try {
      const payload = new URLSearchParams({
        'user_session[email]': this.username,
        'user_session[password]': this.password,
      });

      const response = await fetch(`${BASE_URL}/login`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/x-www-form-urlencoded',
        },
        body: payload.toString(),
        redirect: 'manual',
      });

      // Extract cookies
      const setCookieHeaders = response.headers.getSetCookie?.() || [];
      setCookieHeaders.forEach((cookie) => {
        const [nameValue] = cookie.split(';');
        const [name, value] = nameValue.split('=');
        if (name && value) {
          this.cookies.set(name.trim(), value.trim());
        }
      });

      // Follow redirect
      const location = response.headers.get('location');
      if (location) {
        const finalResponse = await fetch(
          location.startsWith('http') ? location : `${BASE_URL}${location}`,
          {
            headers: {
              Cookie: this.getCookieHeader(),
            },
          }
        );

        if (finalResponse.ok && !finalResponse.url.endsWith('/login')) {
          this.loggedIn = true;
          return true;
        }
      }

      return false;
    } catch (error) {
      console.error('Login error:', error);
      return false;
    }
  }

  /**
   * Authenticated GET for any ShiftGen page, returning its HTML.
   */
  async fetchPage(path: string): Promise<string | null> {
    if (!this.loggedIn) {
      throw new Error('Not logged in. Call login() first.');
    }

    try {
      const url = path.startsWith('http') ? path : `${BASE_URL}${path}`;
      const response = await fetch(url, {
        headers: {
          Cookie: this.getCookieHeader(),
        },
      });

      if (!response.ok) {
        console.error(`Fetch failed for ${path}: ${response.status}`);
        return null;
      }

      return await response.text();
    } catch (error) {
      console.error(`Fetch error for ${path}:`, error);
      return null;
    }
  }

  /**
   * Fetch the multi-site schedule view, which contains every schedule the
   * authenticated account can see.
   */
  async getMultiSiteSchedule(): Promise<string | null> {
    return this.fetchPage(MULTI_SITE_SCHEDULE_PATH);
  }

  /**
   * Get cookie header string
   */
  private getCookieHeader(): string {
    return Array.from(this.cookies.entries())
      .map(([name, value]) => `${name}=${value}`)
      .join('; ');
  }
}
